### DISCUSSION TITLE
Is it allowed to use notFound() inside layout

### DISCUSSION THREAD

**Discussion by horaklukas** [vercel/next.js#53149] [App Router]
I didn't find a mention about it anywhere in the documentation so I come here for a help:

Is it possible to trigger display not found page with `notFound()` when I'm in the layout? Of course I'm asking because it doesn't work to me and produces

 **`Error: notFound() is not allowed to use in root layout`** 

although I'm triggering the error in nested layout not in the root layout.

The route hiearchy looks like
```
app/
  (public-pages)/
     layout.tsx        👈 root layout
     not-found.tsx
     workspace/
       [slug]/
          layout.tsx.  👈 here the notFound() is triggered
          page.tsx
          people/
            page.tsx
```

there is a fetch of workspace data inside the `workspace/[slug]/layout.tsx` 

```tsx
export const getData = async (workspaceSlug: string) => {
  try {
    const workspace = await getWorkspaceDetail(workspaceSlug);

    return workspace;
  } catch (error) {
    return notFound();
  }
};

export type WorkspaceLayoutProps = {
  params: {
    slug: string;
  };
  children: React.ReactNode;
};

export async function generateMetadata({ params: { slug }}: WorkspaceLayoutProps): Promise<Metadata> {
  const workspace = await getData(slug);

  ...
}

export default async function WorkspacePageLayout({ params: { slug }, children}: WorkspaceLayoutProps) {
  const workspace = await getData(slug);
  
  ...
}
```

**Comment by JurajKavka** [comment]
I found just now, that overcome error *"Error: notFound() is not allowed to use in root layout"* I had to create custom `not-found.tsx` file. But what is worst ... request for "not found" URLs does not returns `404` but `200`. And SSR renders loading page and not 404 page... it is wrong and I'm not sure, it is the issue or I'm doing something wrong... but anyway I'm considering to switch completely back to `pages` router. App router has also issue with permanent redirect for example.

**Reply by JurajKavka** [reply]
My version is now `13.4.12` and notFound() returns `200`. I tried `13.4.0` with the same result. But in docs [here](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming) i found: 

*"Next.js will wait for data fetching inside [generateMetadata](https://nextjs.org/docs/app/api-reference/functions/generate-metadata) to complete before streaming UI to the client. This guarantees the first part of a streamed response includes <head> tags."*

So when I make request on blog post based on *slug* in `generateMetadata` and blog post is not found, I raise `notFound()` and it properly responds to browser with `HTTP 404` error.

**Comment by TonsiTT** [comment]
Facing the same issue. It was working fine on old versions of Next.js, but after recent updates, it started giving the error "Error: notFound() is not allowed to use in root layout"

**Comment by horaklukas** [comment]
Ok, apparently it's a real issue - https://github.com/vercel/next.js/issues/53371
